[DSYS-3301]: add Chat Composer documentation website page (#2852)
twilio-labs/paste: packages/paste-website/src/pages/components/chat-composer/index.mdx

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/chat-composer"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(frontmatter: {slug: {eq: "/components/chat-composer/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
        description
      }
      headings {
        depth
        value
      }
    }
    allAirtable(filter: {data: {Feature: {eq: "Chat Composer"}}}) {
      edges {
        node {
          data {
            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
            status
          }
        }
      }
    }
  }
`;

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/chat-composer"
  storybookUrl="/?path=/story/components-chat-composer--default"
  data={props.data}
  description={props.data.mdx.frontmatter.description}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

> <LivePreview

Reviewer: suggestion: the page's first example usually shows the component in isolation, so I think this should be the composer with some initial or placeholder text. 

This chat dialog example w/ the composer in it would be more useful as one of the examples. We should add it there with some explanation text on how it's meant to be used there, and include a link to the Chat Log docs page.

  noInline
  language="jsx"
  scope={{
    Box,
    ChatLog,
    ChatBookend,
    ChatBookendItem,
    ChatEvent,
    ChatMessage,
    ChatBubble,
    ChatMessageMeta,
    Avatar,
    ChatAttachment,
    ChatAttachmentDescription,
    ChatAttachmentLink,
    ChatMessageMetaItem,
    ChatComposer,
    AutoScrollPlugin,
    Button,
    SendIcon,
  }}
>
  {ChatDialogExample}
</LivePreview>

## Guidelines

### About Chat Composer

> A Chat Composer is an input made for users to type rich chat messages. Chat Composer can be used as one part of larger chat user interfaces to provide a seamless authoring experience.

Reviewer: suggestion: we should add a sentence here about how the most common use case in Paste is to use it along with the components in Chat log (and link those docs) to create a full chat experience.

### Accessibility

> Chat Composer supports a variety of aria attributes which are passed into the content editable region of the component.

Reviewer: suggestion: this is vague. Are there any aria attributes we want to suggest? We use aria-label in all the examples, it would be useful to detail that here. Accessibility is ultimately up to developers but this part of the docs is a good place to steer them in the right direction.

### Lexical and Plugins

Chat Composer is built ontop of [Lexical](https://lexical.dev/). The editor itself is extensible and follows a declarative approach to configuration via JSX. Developers can leverage a

> wide variety of existing plugins or write their own to augment the behavior of Chat Composer. To include new Chat Composer behavior provide plugins via the `children` prop.

Reviewer: suggestion: we've made something pretty powerful in its composability and flexibility, and I think we should be more explicit about that by listing out the plugins that we currently baked into the component, and including some examples of how they can build on top of it with other useful plugins

## Examples

### Basic

> 

Reviewer: issue: Each of these examples should have some explanation text above them, since we don't know what level of familiarity someone coming to the docs has with our system or coding in general (many times its designers). It's a little hand-hold-y but I mean, it's supposed to be. For example, for this one I'd point out that the config prop is an object and explain what you can put into it for it to work in a basic way. For the Placeholder one, saying that you use a placeholder prop that will render the string as placeholder text, and so on.

Nitpick: I thiiiiink it'd be better to use a more realistic namespace than 'foo', like 'customer chat'

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With Placeholder

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With Initial Value

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} initialValue="This is my initial value" placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With Max Height

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} maxHeight="size10" ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

> 

Reviewer: todo: let's include an example where we use a lexical plugin in composition with the component, like the scrolling example Lee showed us. It's such a cool thing that this component can do, melds right into the principles of our system. Let's show it off!!

Reviewer 2: I think it'd be cool to show how one would extend this component if Lee already has an easy to port example built. If building from scratch, we can omit until it comes up. But I think this component/lexical by nature is something that is a composed UI for different products, so showcasing that here could be powerful.

Reviewer 3: I think theres a markdown formatter you could use? That one is fun... or they have a TON of example plugins here: https://github.com/facebook/lexical/tree/main/packages/lexical-playground/src/plugins

## Usage Guide

### API

#### Installation

```bash
yarn add @twilio-paste/chat-composer - or - yarn add @twilio-paste/core
```

#### Usage

```jsx
import {ChatComposer} from '@twilio-paste/core/chat-composer';

export const BasicChatComposer = () => (
  <ChatComposer
    config={{
      namespace: 'foo',
      onError: (e) => {
        throw e;
      },
    }}
    ariaLabel="A basic chat composer"
    placeholder="Type here"
  />
);
```

#### Props

| Prop            | Type                                                        | Description                                                                               | Default                                   |
| --------------- | ----------------------------------------------------------- | ----------------------------------------------------------------------------------------- | ----------------------------------------- | --- |

> | `config`        | `LexicalComposerProps['initialConfig']`                     | Configuration provided to the Lexical Composer                                            |                                           |

Reviewer: question: do we want to specify that these are undefined by default or is leaving that field in the table blank enough.

| `children?`     | `LexicalComposerProps['children]`                           | Children provided to the Lexical Composer (i.e. Lexical plugins)                          |                                           |
| `element?`      | `string`                                                    | Overrides the default element name to apply unique styles with the Customization Provider | `'CHAT_COMPOSER'`                         |

> | `placeholder?`  | `string                                                     | JSX.Element`                                                                              | Placeholder content for the Chat Composer |     |

Reviewer: ```suggestion
| `placeholder?`  | `string` or JSX.Element`                   | Placeholder content for the Chat Composer |     |
```

| `onChange?`     | `(editorState: EditorState, editor: LexicalEditor) => void` | Callback invoked when changes occurs within the Chat Composer                             |                                           |
| `initialValue?` | `string`                                                    | Initial value for the Chat Composer                                                       |                                           |
| `maxHeight?`    | `string`                                                    | Maximum height for the Chat Composer                                                      |                                           |

### Figma

<Callout variant="neutral" marginY="space70">
  <CalloutHeading as="span">Figma usage guidelines coming soon</CalloutHeading>
  <CalloutText>
    You can find the Chat Composer components in{' '}
    <Anchor
      href="https://www.figma.com/file/E6KUvMhioUmAgN0nwmReTM/Paste-Components?node-id=15140%3A99225&t=mNw9lFc5iiTDSIZ7-0"
      showExternal
    >
      the Paste Components library
    </Anchor>

>     .

Reviewer: ```suggestion
  
```

    }
  }
`;

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/chat-composer"
  storybookUrl="/?path=/story/components-chat-composer--default"
  data={props.data}
  description={props.data.mdx.frontmatter.description}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

## Guidelines

### About Chat Composer

A Chat Composer is an input made for users to type rich chat messages. Chat Composer is best used as one part of larger chat user interface to provide a seamless authoring experience.
Within the context of Paste, Chat Composer is most typically used alongside the [`Chat Log`](/components/chat-log) component.

### Accessibility

Chat Composer supports a variety of aria attributes which are passed into the content editable region of the component. It is strongly encouraged to use an `aria-label` to describe the
input itself.

> ### Lexical and Plugins

Reviewer 4: question: is there a Lexical plugin to show the rich text toolbar? there's quite a few chat products that use that like Flex. it would be a huge win to show a live example of how to compose that with Lexical plugin if one exists

> Chat Composer is built ontop of the [Lexical](https://lexical.dev) editor. Lexical is extensible and follows a declarative approach to configuration via JSX. Developers can leverage a

Reviewer 4: ```suggestion
Chat Composer is built on top of the [Lexical](https://lexical.dev) editor. Lexical is extensible and follows a declarative approach to configuration via JSX. Developers can leverage a
```

wide variety of existing plugins or write their own to augment the behavior of Chat Composer. Plugins are provided via the `children` prop. As an example,
Chat Composer uses a custom plugin [`AutoLinkPlugin`](https://github.com/twilio-labs/paste/blob/main/packages/paste-core/components/chat-composer/src/AutoLinkPlugin.tsx) internally
which you can see being configured [here](https://github.com/twilio-labs/paste/blob/main/packages/paste-core/components/chat-composer/src/ChatComposer.tsx#L116) as a JSX child.

## Examples

> ### Basic

Reviewer 4: question: do we need to include this live example? are we recommending that a chat composer would be used without placeholder text?

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

> ### With Placeholder

Reviewer 4: todo(blocking): swap headings in the Examples section to be in sentence case

Set a placeholder value using a `placeholder` prop.

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

> 

Reviewer 4: idea: could we add a live example of a chat composer with rich text inside? something to show that it's possible. my quick put together:

Hello **world!** This is a _chat composer_ with rich text functionality.

### With Initial Value

Set an initial value using an `initialValue` prop.

<LivePreview scope={{ChatComposer}} language="jsx">

>   {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} initialValue="This is my initial value" placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}

Reviewer 4: issue(blocking): I noticed that in a few of the examples, the placeholder is `Type Here...`. could we change it to match the designs to say `Chat text`? I don't feel strongly about `Chat text` as being a great placeholder either, would just like to be consistent in both place. 

At a minimum, we could swap to [sentence case](https://paste.twilio.design/foundations/content/product-style-guide/#sentence-case) with `Type here...`

</LivePreview>

### With Max Height

Restrict the height of the composer using a `maxHeight` prop.

> <LivePreview scope={{ChatComposer}} language="jsx">

Reviewer 4: suggestion: can we add some dummy text in the chat composer here so there's a visual cue of the max height in action? at first glance, it just looks like a big white box since the input borders are removed here as well

  {`<ChatComposer config={{namespace: 'foo', onError: (e) => { throw e } }} maxHeight="size10" ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

> ### With ChatLog

Reviewer 4: suggestion: naming this after the component instead of the React component would be nice since this example would also apply to Figma

```suggestion
### With Chat Log
```

> Use Chat Composer alongside other Paste components such as [Chat Log](/components/chat-log) to build more complex chat UI.

Reviewer 4: suggestion: can we add some guidance here about the `AutoScrollPlugin`? I see it used in the live example but unsure if it's clear to consumers why it's there and how to use it

## Examples

### With placeholder

Set a placeholder value using a `placeholder` prop.

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'customer-chat', onError: (e) => { throw e } }} placeholder="Chat text" ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With initial value

Set an initial value using an `initialValue` prop.

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'customer-chat', onError: (e) => { throw e } }} initialValue="This is my initial value" placeholder="Type Here..." ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With max height

Restrict the height of the composer using a `maxHeight` prop.

<LivePreview scope={{ChatComposer, tallInitialText}} language="jsx">
  {`<ChatComposer config={{editorState: tallInitialText, namespace: 'customer-chat', onError: (e) => { throw e } }} maxHeight="size10" ariaLabel="A Basic Chat Composer" />`}
</LivePreview>

### With rich text

Set a rich text value using one of the Lexical formatting APIs such as [`toggleFormat`](https://lexical.dev/docs/api/classes/lexical.TextNode#toggleformat)

> <LivePreview scope={{ChatComposer, richInitialText}} language="jsx">

Reviewer 2: suggestion: I like how clean the codeblocks are, but it may be unclear if folks can't see the `richInitialText` code in this sample. Any chance we can inline it?

            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
            status
          }
        }
      }
    }
  }
`;

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/chat-composer"
  storybookUrl="/?path=/story/components-chat-composer--default"
  data={props.data}
  description={props.data.mdx.frontmatter.description}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'customer-chat', onError: (e) => { throw e } }} placeholder="Chat text" ariaLabel="A basic chat composer" />`}
</LivePreview>

## Guidelines

### About Chat Composer

A Chat Composer is an input made for users to type rich chat messages. Chat Composer is best used as one part of larger chat user interface to provide a seamless authoring experience.
Within the context of Paste, Chat Composer is most typically used alongside the [`Chat Log`](/components/chat-log) component.

### Accessibility

> Chat Composer supports a variety of aria attributes which are passed into the content editable region of the component. If the surrounding UI includes a screen reader visible label

Reviewer 3: Suggestion(non-blocking): It might be easier to read if these if statements were in a bulleted list.

wide variety of [existing plugins](https://github.com/twilio-labs/paste/blob/main/packages/paste-libraries/lexical/src/index.tsx) via the `@twilio-paste/lexical-library` package or other
sources. Alternatively, developers can write their own custom plugin logic. Plugins are provided to the Chat Composer via the `children` prop.

#### Auto Link Plugin

Chat Composer uses a custom [`AutoLinkPlugin`](https://github.com/twilio-labs/paste/blob/main/packages/paste-core/components/chat-composer/src/AutoLinkPlugin.tsx) internally
which you can see being configured [here](https://github.com/twilio-labs/paste/blob/main/packages/paste-core/components/chat-composer/src/ChatComposer.tsx#L116) as a JSX child.

## Examples

### With placeholder

Set a placeholder value using a `placeholder` prop.

<LivePreview scope={{ChatComposer}} language="jsx">
  {`<ChatComposer config={{namespace: 'customer-chat', onError: (e) => { throw e } }} placeholder="Chat text" ariaLabel="A placeholder chat composer" />`}
</LivePreview>

### With initial value

> Set an initial value using an `initialValue` prop.

Reviewer 3: Suggestion: I think it would be good to say that the initialValue prop only supports having a single line string. If they want anything more custom than that, they need to set the editorState in the config prop.
